### ISSUE TITLE
[BUG] Volume expansion and snapshot deletion leading to node overfill and faulted s

### ISSUE THREAD

**Issue by jani888** [longhorn/longhorn#9565]
## Describe the bug
We are encountering two issues related to volume expansion and snapshot deletion:

When expanding a volume to match the size of the node, the volume occasionally fills the entire node, causing it to go into a faulted state.
While attempting to delete a snapshot from a 3Ti volume (with 3 existing snapshots), Longhorn filled the entire node, putting the volume into a faulted state. Initially, the node had 4Ti of free space. After expanding the node to 5Ti, the volume became healthy again, but the snapshot deletion process seems to continue consuming additional space, potentially filling the node again.
We need clarification on whether this is expected behavior and how we can prevent or mitigate it in the future.

## To Reproduce
Steps to reproduce the issue:

1. Create a volume of 3Ti with snapshots.
2. Expand the volume to match the size of a node.
3. Attempt to delete one of the snapshots.
4. Observe that Longhorn fills the available space on the node, leading to a faulted state.

## Expected behavior
- Volume expansion should only allocate the requested size and should not overfill the node.
- Snapshot deletions should not cause Longhorn to fill the entire disk during the process.

## Support bundle for troubleshooting
- Support bundle will be provided via email.

## Environment
- Longhorn version: v1.7.0
- Impacted volume (PV): pvc-3ec28539-0d27-4122-929d-0302207b81ba, pvc-4d48e3ed-7d3c-440b-83bb-92c951685a21
- Installation method: Helm
 - Kubernetes distro and version: RKE2 v1.28.11+rke2r1
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 39
 - Node config
   - OS type and version: 
   - Kernel version: 
   - CPU per node: 30-45 core
   - Memory per node: 64-256 Gi
   - Disk type: SSD and HDD
   - Network bandwidth between the nodes: 10 Gbps
 - Underlying Infrastructure: Baremetal
 - Number of Longhorn volumes in the cluster: 39

## Additional context
The snapshot deletion issue seems to be consuming more space than expected. We want to know if this behavior is normal and how we can prepare the system to avoid filling the disk during future snapshot deletions or volume expansions.

## Workaround and Mitigation
We increased the size of the node to 5Ti, which temporarily made the volume healthy again, but the delete process continues to consume space.

**Comment by jani888** [comment]
I understand this, however I don't think it's a good idea, to delete snapshots after a volume expansion, if there is not enough space for it. This way there is a serious risk of data loss, because when if deletes the snapshot, the node will fill over, and the volume goes into a faulted state. So with a volume expansion we can lose the entire volume. Is it possible to implement a safety check, and surpress the snapshot deletion until there is guaranteed to be enough space for it on the node?

**Comment by c3y1huang** [comment]
> I understand this, however I don't think it's a good idea, to delete snapshots after a volume expansion, if there is not enough space for it. This way there is a serious risk of data loss, because when if deletes the snapshot, the node will fill over, and the volume goes into a faulted state. So with a volume expansion we can lose the entire volume. Is it possible to implement a safety check, and surpress the snapshot deletion until there is guaranteed to be enough space for it on the node?

I think this is a fair suggestion to implement a safe check on the available space before deleting a snapshot. What do you think, @shuo-wu @PhanLe1010 @derekbit ?

**Comment by derekbit** [comment]
> > I understand this, however I don't think it's a good idea, to delete snapshots after a volume expansion, if there is not enough space for it. This way there is a serious risk of data loss, because when if deletes the snapshot, the node will fill over, and the volume goes into a faulted state. So with a volume expansion we can lose the entire volume. Is it possible to implement a safety check, and surpress the snapshot deletion until there is guaranteed to be enough space for it on the node?
> 
> I think this is a fair suggestion to implement a safe check on the available space before deleting a snapshot. What do you think, @shuo-wu @PhanLe1010 @derekbit ?

Agree with the idea.

**Comment by shuo-wu** [comment]
I am fine with adding a check. But we need to consider some special cases:
1. System snapshot auto cleanup. Should we skip this (for rebuilding or recurring jobs) if there is no enough space?
2. The parent of the head deletion. For this one, Longhorn does not need any extra space
3. v2 snapshot deletion. IIUC, SPDK does not need any extra space either

**Comment by jani888** [comment]
Hi @derekbit, @c3y1huang, @shuo-wu,

Just checking in -- any updates on this issue?

Thanks!

**Comment by jani888** [comment]
Hi @derekbit, @c3y1huang, @shuo-wu,

I wanted to follow up on this issue, as I haven't seen any updates since my last comment. Could you provide any progress details or confirm if this will be addressed in the v1.9.0 milestone?

The lack of safeguards against node overfilling during snapshot deletion remains a significant risk for our production environment. It would be helpful to know if the proposed space-check mechanism or any other mitigation steps are being actively worked on.

Looking forward to your response.

Thanks!

**Comment by derekbit** [comment]
@jani888 Yes, the issue is scheduled to v1.9.0. Thanks for your feed back.
